src/main/scala/beam/utils/map/ActivitiesClustering.scala

```diff
@@ -0,0 +1,136 @@
+
+private case class ClusterAttributes(size: Int, homes: Int, works: Int) extends Attributes
+
+class ActivitiesClustering(val pathToPlansCsv: String, nClusters: Int) extends StrictLogging {
+  private val geometryFactory: GeometryFactory = new GeometryFactory()
+  private val geoUtils = new GeoUtils {
+    def localCRS: String = "epsg:26910"
+  }
+
+  def run(): Unit = {
+    val clusters = cluster()
+    val clusterWithConvexHull = clusters.map { c =>
+      val coords = c.activitiesLocation.map { act =>
+        new Coordinate(act.coord.getX, act.coord.getY)
+      }
+      val convexHullGeom = new ConvexHull(coords.toArray, geometryFactory).getConvexHull
+      c -> convexHullGeom
+    }
+    val shapeWriter = ShapeWriter.worldGeodetic[Polygon, ClusterAttributes]("clusters.shp")
+    clusterWithConvexHull.zipWithIndex.foreach {
+      case ((c, geom), idx) =>
+        if (geom.getNumPoints > 2) {
+          val wgsCoords = geom.getCoordinates.map { c =>
+            val wgsCoord = geoUtils.utm2Wgs(new Coord(c.x, c.y))
+            new Coordinate(wgsCoord.getX, wgsCoord.getY)
+          }
+          try {
+            val polygon = geometryFactory.createPolygon(wgsCoords)
+            val nHomes = c.activitiesLocation.count(x => x.label == "Home")
+            val nWorks = c.activitiesLocation.count(x => x.label == "Work")
+            shapeWriter.add(
+              polygon,
+              idx.toString,
+              ClusterAttributes(size = c.size, homes = nHomes, works = nWorks)
+            )
+          } catch {
+            case NonFatal(ex) =>
+              logger.error("Can't create or add", ex)
+          }
+        }
+    }
+    shapeWriter.write()
+  }
+
+  private def cluster(): Array[ClusterInfo] = {
+    val activities = readActivities(pathToPlansCsv)
+    val db = createDatabase(activities)
+    val kmeans = new KMeansElkan[NumberVector](
+      SquaredEuclideanDistanceFunction.STATIC,
+      nClusters,
+      1000,
+      new RandomUniformGeneratedInitialMeans(RandomFactory.DEFAULT),
+      true
+    )
+    val result = kmeans.run(db)
+    result.getAllClusters.asScala.zipWithIndex.map {
+      case (cluster, idx) =>
+        logger.info(s"# $idx: ${cluster.getNameAutomatic}")
+        logger.info(s"Size: ${cluster.size()}")
+        logger.info(s"Model: ${cluster.getModel}")
+        logger.info(s"Center: ${cluster.getModel.getMean.toVector}")
+        logger.info(s"getPrototype: ${cluster.getModel.getPrototype.toString}")
+        val vectors = db.getRelation(TypeUtil.DOUBLE_VECTOR_FIELD)
+        val labels = db.getRelation(TypeUtil.STRING)
+        val coords: ArrayBuffer[CoordWithLabel] = new ArrayBuffer(cluster.size())
+        val iter: DBIDIter = cluster.getIDs.iter()
+        while (iter.valid()) {
+          val o: DoubleVector = vectors.get(iter)
+          val arr = o.toArray
+          val coord = new Coord(arr(0), arr(1))
+          coords += CoordWithLabel(coord, labels.get(iter))
```

Comment:
It's still a buffer, not an array, so you can't do an update on random index:
```scala
val arrBuf = collection.mutable.ArrayBuffer[Double](5)
arrBuf(0) = 1.0
arrBuf(1) = 2.0
```
You will get `java.lang.IndexOutOfBoundsException:`
![image](https://user-images.githubusercontent.com/5107562/80002545-f9b73000-84e9-11ea-9f3c-9fea44614055.png)
Pre-allocation of ArrayBuffer is just not to re-create internal array if there is no enough capacity

Reply:
You alrady know the size in advance: ` val coords: ArrayBuffer[CoordWithLabel] = new ArrayBuffer(cluster.size())`. Maybe would be better to use some `Indexed` datastructure to avoid overhead